sys/net/network_layer/ng_sixlowpan/frag/rbuf.c

```diff
@@ -0,0 +1,307 @@
+
+        entry->ints->start = 0;
+        entry->ints->end = 0;
+        entry->ints->next = NULL;
+        entry->ints = next;
+    }
+
+    entry->pkt = NULL;
+}
+
+static bool _rbuf_update_ints(rbuf_t *entry, uint16_t offset, size_t frag_size)
+{
+    rbuf_int_t *ptr = entry->ints, *new;
+    uint16_t end = (uint16_t)(offset + frag_size - 1);
+
+    while (ptr != NULL) {
+        if (_rbuf_int_in(ptr, offset, end)) {
+            DEBUG("6lo rfrag: overlapping or same intervals.\n");
+            return false;
+        }
+
+        ptr = ptr->next;
+    }
+
+    new = _rbuf_int_get_free();
+
+    if (new == NULL) {
+        DEBUG("6lo rfrag: no space left in rbuf interval buffer.\n");
+        return false;
+    }
+
+    new->start = offset;
+    new->end = end;
+
+    DEBUG("6lo rfrag: add interval (%" PRIu16 ", %" PRIu16 ") to entry (%s, ",
+          new->start, new->end, ng_netif_addr_to_str(l2addr_str,
+                  sizeof(l2addr_str), entry->src, entry->src_len));
+    DEBUG("%s, %zu, %" PRIu16 ")\n", ng_netif_addr_to_str(l2addr_str,
+            sizeof(l2addr_str), entry->dst, entry->dst_len), entry->pkt->size,
+          entry->tag);
+
+    LL_PREPEND(entry->ints, new);
+
+    return true;
+}
+
+static void _rbuf_gc(void)
+{
+    rbuf_t *oldest = NULL;
+    timex_t now, timeout = RBUF_TIMEOUT;
+    int i;
+
+    vtimer_now(&now);
+
+    for (i = 0; i < RBUF_SIZE; i++) {
+        if ((rbuf[i].pkt != NULL) &&
+            (timex_cmp(timex_sub(now, rbuf[i].arrival), timeout) >= 0)) {
+            DEBUG("6lo rfrag: entry (%s, ", ng_netif_addr_to_str(l2addr_str,
+                    sizeof(l2addr_str), rbuf[i].src, rbuf[i].src_len));
+            DEBUG("%s, %zu, %" PRIu16 ") timed out\n",
+                  ng_netif_addr_to_str(l2addr_str, sizeof(l2addr_str), rbuf[i].dst,
+                                       rbuf[i].dst_len),
+                  rbuf[i].pkt->size, rbuf[i].tag);
+
+            ng_pktbuf_release(rbuf[i].pkt);
+            _rbuf_rem(rbuf + i);
+        }
+        else if ((oldest == NULL) || (timex_cmp(rbuf[i].arrival, oldest->arrival))) {
+            oldest = rbuf + i;
+        }
+    }
+
+    if (((i >= RBUF_SIZE) && (oldest != NULL))) {
+        DEBUG("6lo rfrag: reassembly buffer full, remove oldest entry");
+        ng_pktbuf_release(oldest->pkt);
+        _rbuf_rem(oldest);
+    }
+}
+
+static rbuf_t *_rbuf_get(const void *src, size_t src_len,
+                         const void *dst, size_t dst_len,
+                         size_t size, uint16_t tag)
+{
+    rbuf_t *res = NULL;
+
+    for (int i = 0; i < RBUF_SIZE; i++) {
+        /* check first if entry already available */
+        if ((rbuf[i].pkt != NULL) && (rbuf[i].pkt->size == size) &&
+            (rbuf[i].tag == tag) && (rbuf[i].src_len == src_len) &&
+            (rbuf[i].dst_len == dst_len) &&
+            (memcmp(rbuf[i].src, src, src_len) == 0) &&
+            (memcmp(rbuf[i].dst, dst, dst_len) == 0)) {
```

Comment:
I don't know how this work, but maybe I will look into it when I found some time. The expensiveness however can't really be discussed, since those are the identifying parameters of a fragment that [SHALL](https://tools.ietf.org/html/rfc2119) be used via [the specification](https://tools.ietf.org/html/rfc4944#section-5.3).